https://issues.redhat.com/browse/ACM-8791 Install known issue (#5904)
open-cluster-management/rhacm-docs: release_notes/known_issues_install.adoc

For more about deprecations and removals, see xref:../release_notes/deprecate_remove.adoc#deprecations-removals[Deprecations and removals].

[#ocm-controller-crash]
== ocm-controller crashes after {product-title-short} upgrade

> After you upgrade from 2.7.x to 2.8.x and then to 2.9.0, the `ocm-controller` of the `multicluster-engine` namespace crashes. The following error message appears after you run the command, `oc get managedclustersets`:

Reviewer: When this issue happens, either ManagedClusterSet or ManagedClusterSetBinding or both are impact. So it's not enough to list only managedclustersets with this command.

> After you upgrade from 2.7.x to 2.8.x and then to 2.9.0, the `ocm-controller` of the `multicluster-engine` namespace crashes. The following error message appears after you run the command, `oc get managedclustersets`:

Author: @elgnay thanks for clarifying. Here is my suggestion 
```suggestion
After you upgrade from 2.7.x to 2.8.x and then to 2.9.0, the `ocm-controller` of the `multicluster-engine` namespace crashes. After you attempt to list and or both, `ManagedClusterSet` and `ManagedClusterSetBinding` custom resource definitions, the following error message appears:
```

Reviewer: It looks good to me.

[source,bash]
----
Error from server: request to convert CR from an invalid group/version: cluster.open-cluster-management.io/v1beta1
----

This indicates that the migration of `ManagedClusterSets` and `ManagedClusterSetBindings` custom resource definitions from `v1beta1` to `v1beta2` failed. 

To resolve this, you must initiate the API migration manually. Completet the following steps:

> //I have concerns about the multiple commands, I see about 

Author: ```suggestion

//I have concerns about the multiple commands, I see about 
```

For more about deprecations and removals, see xref:../release_notes/deprecate_remove.adoc#deprecations-removals[Deprecations and removals].

[#ocm-controller-crash]

> == ocm-controller crashes after {product-title-short} upgrade

Reviewer 2: According to modular documentation, if it's a task/procedure module/doc, then we ought to have the title be a gerund verb. 

What do you think about something like:

== Troubleshooting errors when ocm-controller crashes after {prod-title-short} upgrade

Author: Good point @jc-berger. I recommend that we move this entire section to the troubleshooting section instead of listing it in known issues.

After you upgrade from 2.7.x to 2.8.x and then to 2.9.0, the `ocm-controller` of the `multicluster-engine` namespace crashes. After you attempt to list and or both, `ManagedClusterSet` and `ManagedClusterSetBinding` custom resource definitions, the following error message appears:

[source,bash]
----
Error from server: request to convert CR from an invalid group/version: cluster.open-cluster-management.io/v1beta1
----

> This indicates that the migration of `ManagedClusterSets` and `ManagedClusterSetBindings` custom resource definitions from `v1beta1` to `v1beta2` failed. 

Reviewer 2: Minor suggestions. 

`this` is vague. Although it's more word count, what do you  think of, `This message indicates that the migration...`

> To resolve this, you must initiate the API migration manually. Completet the following steps:

Reviewer 2: Same as above. How about, `To resolve this error, initiate the API migration manually...`

> //I have concerns about the multiple commands, I see about 

Author: ```suggestion
//I have concerns about the multiple commands. I recommend creating substeps so that the user understands the purpose for each command. 
```

> //I have concerns about the multiple commands. I recommend creating substeps so that the user understands the purpose for each command. 

Author: @elgnay

> //I have concerns about the multiple commands. I recommend creating substeps so that the user understands the purpose for each command. 

Reviewer 2: Reading the substeps below, I think they read well!

. Revert the `cluster-manager` to a previous release. Run the following commands:

[source,bash]
----
oc annotate mce multiclusterengine pause=true
oc patch deployment cluster-manager -n multicluster-engine -p \  '{"spec":{"template":{"spec":{"containers":[{"name":"registration-operator","image":"registry.redhat.io/multicluster-engine/registration-operator-rhel8@sha256:35999c3a1022d908b6fe30aa9b85878e666392dbbd685e9f3edcb83e3336d19f"}]}}}}'
export ORIGIN_REGISTRATION_IMAGE=$(oc get clustermanager cluster-manager -o jsonpath='{.spec.registrationImagePullSpec}')
oc patch clustermanager cluster-manager --type='json' -p='[{"op": "replace", "path": "/spec/registrationImagePullSpec", "value": "registry.redhat.io/multicluster-engine/registration-rhel8@sha256:a3c22aa4326859d75986bf24322068f0aff2103cccc06e1001faaf79b9390515"}]' 
----

> 

Author: I'm not sure the purpose for each one, just making an attempt at the substep suggestion. For example:

```suggestion
. Revert the `cluster-manager` to a previous release. 
.. Pause the `multiclusterengine` with the following command:
+
[source,bash]
----
oc annotate mce multiclusterengine pause=true
----

.. Run the following command to update the registration image for the `cluster-manage` deployment:
+
[source,bash]
----
oc patch deployment cluster-manager -n multicluster-engine -p \  '{"spec":{"template":{"spec":{"containers":[{"name":"registration-operator","image":"registry.redhat.io/multicluster-engine/registration-operator-rhel8@sha256:35999c3a1022d908b6fe30aa9b85878e666392dbbd685e9f3edcb83e3336d19f"}]}}}}'
export ORIGIN_REGISTRATION_IMAGE=$(oc get clustermanager cluster-manager -o jsonpath='{.spec.registrationImagePullSpec}')
----

.. Replace the registry path of the `cluster-manager` deployment with the previous registration. Run the following command:
+
[source,bash]
----
oc patch clustermanager cluster-manager --type='json' -p='[{"op": "replace", "path": "/spec/registrationImagePullSpec", "value": "registry.redhat.io/multicluster-engine/registration-rhel8@sha256:a3c22aa4326859d75986bf24322068f0aff2103cccc06e1001faaf79b9390515"}]' 
----

```

Reviewer: > .. Run the following command to update the registration image for the `cluster-manage` deployment:

.. Run the following command to replace the image of the `cluster-manager` deployment with a previous version:

> .. Replace the registry path of the `cluster-manager` deployment with the previous registration. Run the following command:

.. Replace the  registration image reference in ClusterManager with a previous version. Run the following command:

oc patch StorageVersionMigration managedclustersetbindings.cluster.open-cluster-management.io --type='json' -p='[{"op":"replace", "path":"/spec/resource/version", "value":"v1beta1"}]'
oc patch StorageVersionMigration managedclustersetbindings.cluster.open-cluster-management.io --type='json' --subresource status -p='[{"op":"remove", "path":"/status/conditions"}]' 
----

. Run the following command to wait for the migration to complete:

[source,bash]
----
oc wait storageversionmigration managedclustersets.cluster.open-cluster-management.io --for=condition=Succeeded --timeout=120s 
oc wait storageversionmigration managedclustersetbindings.cluster.open-cluster-management.io --for=condition=Succeeded --timeout=120s
----

. Restore the `cluster-manager` back to {product-title-short} {product-version}. It might take several minutes. Run the following command:

[source,bash]
----
oc annotate mce multiclusterengine pause-
oc patch clustermanager cluster-manager --type='json' -p='[{"op": "replace", "path": "/spec/registrationImagePullSpec", "value": "'$ORIGIN_REGISTRATION_IMAGE'"}]' 
----

> . Run the following commands to verify that {product-title-short} is recovered:

Reviewer 2: Typically, in modular documentation, verification steps have its own section. So, it would look like:

=== Verification or .Verification 

- To verify that {prod-title-shot} is recovered, run the following command:

Note, IBM SG encourages (though not strictly requires) to lead with reason before action. So, the reason is to verify, and the action is to run the command.

[source,bash]
----
oc get managedclusterset
oc get managedclustersetbinding -A 
----

> //what will the user see?

Author: ```suggestion
//what will the user see after running the commands? 
```

> //what will the user see after running the commands? 

Reviewer 2: Good point, even something simple like, 

- View the cluster set and confirm it meets your needs. 

With that said, I don't think we need to hold off from merging for this reason.

Reviewer: After running the commands, the ManagedClusterSets and ManagedClusterSetBindings should be listed without any error message.
